File: 222_Count_Complete_Tree_Nodes.c
// 2015-06-06  Runtime: 55 ms
// thanks to https://leetcode.com/discuss/38899/easy-short-c-recursive-solution
// Python总是超时, 试了二分法求最底层的叶节点个数还是超时 (Special Judge: very large tree) 😢

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int countNodes(struct TreeNode* root) {
    if (!root) return 0;
    int leftHeight = 0, rightHeight = 0;
    struct TreeNode *L = root, *R = root;
    while(L)
        L = L -> left, leftHeight++;
    while(R)
        R = R -> right, rightHeight++;
    if (leftHeight == rightHeight) return (1 << rightHeight) - 1; // 2 ^ rightHeight - 1
    return 1 + countNodes(root -> left) + countNodes(root -> right);
}

File: 222_Count_Complete_Tree_Nodes.py
# 2015-11-10  Runtime: 164 ms
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root: return 0
        if not root.left and not root.right: return 1
        p1, p2, L, R = root.left, root.right, 0, 0
        while p1:
            p1, L = p1.left, L + 1
        while p2:
            p2, R = p2.left, R + 1
        if L == R:
            return 1 + 2 ** L - 1 + self.countNodes(root.right)
        else:
            return 1 + 2 ** R - 1 + self.countNodes(root.left)